runtime/gc_glue_java/MetronomeDelegate.cpp

```diff
@@ -1115,5 +1133,396 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+	 *
+	 * It is also important to scan JNI global references after scanning threads, because
+	 * the JNI global reference barrier is executed at deletion time, not creation time.
+	 * We could have barriers in both, but controlling the order is an easy solution.
+	 *
+	 *
+	 * The Metronome write barrier ensures that no unscanned thread can expose an object
+	 * to other threads without it becoming a root and no scanned thread can make an
+	 * object that was once reachable unreachable until it has been traced ("snapshot at
+	 * the beginning with a fuzzy snapshot").  This eliminates other order dependencies
+	 * between portions of the scan or requirements that multiple phases be done as an
+	 * atomic unit.  However, some phases are still done atomically because we have not
+	 * yet determined whether the iterators that they use are safe and complete and have
+	 * not even analyzed in all cases whether correctness depends on completeness.
+	 */
+	if (env->_currentTask->synchronizeGCThreadsAndReleaseMaster(env, UNIQUE_ID)) {
+#if defined(J9VM_GC_FINALIZATION)
+		/* Note: if iterators are safe in scanFinalizableObjects, disableYield() could be
+		 * removed.
+		 */
+		env->disableYield();
+		rootMarker.scanFinalizableObjects(env);
+		env->enableYield();
+		_scheduler->condYieldFromGC(env);
+#endif /* J9VM_GC_FINALIZATION */
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		if (!isDynamicClassUnloadingEnabled()) {
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+			/* We are scanning all classes, no need to include stack frame references */
+			rootMarker.setIncludeStackFrameClassReferences(false);
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		} else {
+			rootMarker.setIncludeStackFrameClassReferences(true);
+		}
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+		env->_currentTask->releaseSynchronizedGCThreads(env);
+	}
+
+	rootMarker.scanThreads(env);
+
+	if (env->_currentTask->synchronizeGCThreadsAndReleaseMaster(env, UNIQUE_ID)) {
+		_extensions->newThreadAllocationColor = GC_MARK;
+		_realtimeGC->disableDoubleBarrier(env);
+		if (_realtimeGC->verbose(env) >= 3) {
+			rootMarker.reportThreadCount(env);
+		}
+
+		/* Note: if iterators are safe for some or all remaining atomic root categories,
+		 * disableYield() could be removed or moved inside scanAtomicRoots.
+		 */
+		env->disableYield();
+		rootMarker.scanAtomicRoots(env);
+		env->enableYield();
+		rootMarker.scanIncrementalRoots(env);
+	
+		env->_currentTask->releaseSynchronizedGCThreads(env);
+	}
+
+	env->setRootScanner(NULL);
+}
+
+void
+MM_MetronomeDelegate::markLiveObjectsScan(MM_EnvironmentRealtime *env)
+{
+	MM_RealtimeMarkingSchemeRootClearer rootScanner(env, _realtimeGC);
+	env->setRootScanner(&rootScanner);
+	rootScanner.scanClearable(env);
+	env->setRootScanner(NULL);
+	Assert_MM_true(env->getGCEnvironment()->_referenceObjectBuffer->isEmpty());
+}
+
+void
+MM_MetronomeDelegate::flushRememberedSet(MM_EnvironmentRealtime *env)
```

Comment:
This is not remember set. 
This should be renamed to flushReferenceObjects (and possibly called from delegate's markLiveObjectsScan - not sure where exactly markLiveObjectsScan is called from)

Reply:
Moving it to the delegate function would change the order of the code from its original implementation. possibly breaking it. markLiveObjectsScan() is called from the spot you showed me the other day in RealtimeMarkTask.